File: solutions/cpp/0349.cpp
class Solution {
 public:
  vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    vector<int> ans;
    unordered_set<int> nums1Set{begin(nums1), end(nums1)};

    for (const int num : nums2)
      if (nums1Set.erase(num))
        ans.push_back(num);

    return ans;
  }
};

File: solutions/java/0349.java
class Solution {
  public int[] intersection(int[] nums1, int[] nums2) {
    List<Integer> ans = new ArrayList<>();
    Set<Integer> set = new HashSet<>();

    for (int num : nums1)
      set.add(num);

    for (int num : nums2)
      if (set.remove(num))
        ans.add(num);

    return ans.stream().mapToInt(i -> i).toArray();
  }
}